hikari/impl/event_manager_base.py

```diff
@@ -50,46 +56,130 @@
     ConsumerT = typing.Callable[
         [gateway_shard.GatewayShard, data_binding.JSONObject], typing.Coroutine[typing.Any, typing.Any, None]
     ]
-    ListenerMapT = typing.MutableMapping[
+    ListenerMapT = typing.Dict[
         typing.Type[event_manager.EventT_co],
-        typing.MutableSequence[event_manager.CallbackT[event_manager.EventT_co]],
+        typing.List[event_manager.CallbackT[event_manager.EventT_co]],
     ]
     WaiterT = typing.Tuple[event_manager.PredicateT[event_manager.EventT_co], asyncio.Future[event_manager.EventT_co]]
-    WaiterMapT = typing.MutableMapping[
-        typing.Type[event_manager.EventT_co], typing.MutableSet[WaiterT[event_manager.EventT_co]]
+    WaiterMapT = typing.Dict[typing.Type[event_manager.EventT_co], typing.Set[WaiterT[event_manager.EventT_co]]]
+
+    EventManagerBaseT = typing.TypeVar("EventManagerBaseT", bound="EventManagerBase")
+    UnboundMethodT = typing.Callable[
+        [EventManagerBaseT, gateway_shard.GatewayShard, data_binding.JSONObject],
+        typing.Coroutine[typing.Any, typing.Any, None],
+    ]
+    MethodT = typing.Callable[
+        [gateway_shard.GatewayShard, data_binding.JSONObject],
+        typing.Coroutine[typing.Any, typing.Any, None],
     ]
 
 
 def _default_predicate(_: event_manager.EventT_inv) -> bool:
     return True
 
 
+_CACHE_RESOURCE_ATTRIBUTE = "__CACHE_RESOURCE__"
+_EVENT_TYPES_ATTRIBUTE = "__EVENT_TYPES__"
+
+
+def filtered(
+    event_types: typing.Union[typing.Type[base_events.Event], typing.Sequence[typing.Type[base_events.Event]]],
+    cache_resource: config.CacheComponents = config.CacheComponents.NONE,
+    /,
+) -> typing.Callable[[UnboundMethodT[EventManagerBaseT]], UnboundMethodT[EventManagerBaseT]]:
+    """Add metadata to a listener method to indicate when it should be unmarshalled and dispatched."""
+    if isinstance(event_types, typing.Sequence):
+        # dict.fromkeys is used to remove any duplicate entries here
+        event_types = tuple(dict.fromkeys(itertools.chain.from_iterable(e.dispatches() for e in event_types)))
+
+    else:
+        event_types = event_types.dispatches()
+
+    def decorator(method: UnboundMethodT[EventManagerBaseT], /) -> UnboundMethodT[EventManagerBaseT]:
+        setattr(method, _CACHE_RESOURCE_ATTRIBUTE, cache_resource)
+        setattr(method, _EVENT_TYPES_ATTRIBUTE, event_types)
+        return method
+
+    return decorator
+
+
+@attr.frozen()
```

Comment:
```suggestion
@attr.define()
```

Reply:
Frozen is used for it's hashability and only a couple dozen of these are built at start-up so there's no real performance concern about the usage